pkg/api/v1/validation.go

```diff
@@ -1,69 +1,47 @@
+package v1
 
-Licensed under the Apache License, Version 2.0 (the "License");
-you may not use this file except in compliance with the License.
-You may obtain a copy of the License at
-
-    http://www.apache.org/licenses/LICENSE-2.0
-
-Unless required by applicable law or agreed to in writing, software
-distributed under the License is distributed on an "AS IS" BASIS,
-WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-See the License for the specific language governing permissions and
-limitations under the License.
-*/
+import (
 
-package validation
+	//"k8s.io/apimachinery/pkg/util/runtime"
 
-import (
 	"fmt"
 	"sort"
 
-	kapi "k8s.io/kubernetes/pkg/api"
-	unversionedvalidation "k8s.io/kubernetes/pkg/api/unversioned/validation"
-	kapivalidation "k8s.io/kubernetes/pkg/api/validation"
-	batchvalidation "k8s.io/kubernetes/pkg/apis/batch/validation"
-	"k8s.io/kubernetes/pkg/util/sets"
-	"k8s.io/kubernetes/pkg/util/validation/field"
-
-	"github.com/davecgh/go-spew/spew"
-	"github.com/golang/glog"
-	wapi "github.com/sdminonne/workflow-controller/pkg/api"
+	batchv1 "k8s.io/api/batch/v1"
+	batchv2 "k8s.io/api/batch/v2alpha1"
+	api "k8s.io/api/core/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/validation"
+	v1validation "k8s.io/apimachinery/pkg/apis/meta/v1/validation"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/validation/field"
 )
 
-// ValidateWorkflow validates a workflow
-func ValidateWorkflow(workflow *wapi.Workflow) field.ErrorList {
-	s := spew.ConfigState{
-		Indent: " ",
-		// Extra deep spew.
-		DisableMethods: true,
-	}
-	glog.V(6).Infof("Workflow %s\n", s.Sdump(workflow))
-
-	allErrs := kapivalidation.ValidateObjectMeta(&workflow.ObjectMeta, true, kapivalidation.ValidateReplicationControllerName, field.NewPath("metadata"))
+// ValidateWorkflow validates Workflow
+func ValidateWorkflow(workflow *Workflow) field.ErrorList {
+	allErrs := validation.ValidateObjectMeta(&workflow.ObjectMeta, true, validation.NameIsDNSSubdomain, field.NewPath("metadata"))
 	allErrs = append(allErrs, ValidateWorkflowSpec(&(workflow.Spec), field.NewPath("spec"))...)
 	return allErrs
 }
 
-// ValidateWorkflowSpec validate WorkflowSpec
-func ValidateWorkflowSpec(spec *wapi.WorkflowSpec, fieldPath *field.Path) field.ErrorList {
+// ValidateWorkflowSpec validates WorkflowSpec
+func ValidateWorkflowSpec(spec *WorkflowSpec, fieldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
 	if spec.ActiveDeadlineSeconds != nil {
-		allErrs = append(allErrs, kapivalidation.ValidateNonnegativeField(int64(*spec.ActiveDeadlineSeconds), fieldPath.Child("activeDeadlineSeconds"))...)
+		allErrs = append(allErrs, validation.ValidateNonnegativeField(int64(*spec.ActiveDeadlineSeconds), fieldPath.Child("activeDeadlineSeconds"))...)
 	}
 	if spec.Selector == nil {
 		allErrs = append(allErrs, field.Required(fieldPath.Child("selector"), ""))
```

Comment:
maybe you can use `const string`for "selector", "steps", "spec"...
it is used in several places.

Reply:
ok

Reply:
Hmm taking back. Following kube convention I'll keep the strings without const.

Reply:
kube: the source of truth ;)